geo/aws.go

```diff
@@ -0,0 +1,175 @@
+package geo
+
+import (
+	"encoding/json"
+	"errors"
+	"io/ioutil"
+	"log"
+	"net"
+	"os"
+	"path/filepath"
+	"time"
+)
+
+const (
+	AWSGeoUrl  = "https://ip-ranges.amazonaws.com/ip-ranges.json"
+	AWSExpired = 7 * 24 * time.Hour
+	AWSFile    = "aws-ip.json"
+)
+
+type AWSProvider struct {
+	Info     *AWSGeoInfo
+	fileName string
+	url      string
+	expire   time.Duration
+	ready    bool
+}
+
+type AWSGeoInfo struct {
+	SyncToken  string      `json:"syncToken"`
+	CreateDate string      `json:"createDate"`
+	Prefixes   []AWSIPInfo `json:"prefixes"`
+}
+
+type AWSIPInfo struct {
+	IPPrefix           string     `json:"ip_prefix"`
+	Region             string     `json:"region"`
+	Service            string     `json:"service"`
+	NetworkBorderGroup string     `json:"network_border_group"`
+	Subnet             *net.IPNet `json:"-"`
+}
+
+var AWSRegionMapping = map[string]string{
+	"us-east-1":      "US",
+	"us-east-2":      "US",
+	"us-west-1":      "US",
+	"us-west-2":      "US",
+	"af-south-1":     "ZA",
+	"ap-east-1":      "HK",
+	"ap-south-1":     "IN",
+	"ap-northeast-1": "JP",
+	"ap-northeast-2": "KR",
+	"ap-northeast-3": "JP",
+	"ap-southeast-1": "SG",
+	"ap-southeast-2": "AU",
+	"ca-central-1":   "CA",
+	"eu-central-1":   "DE",
+	"eu-west-1":      "IE",
+	"eu-west-2":      "GB",
+	"eu-west-3":      "FR",
+	"eu-south-1":     "IT",
+	"eu-north-1":     "SE",
+	"me-south-1":     "BH",
+	"sa-east-1":      "BR",
+}
+
+func NewAWSProvider(path string) *AWSProvider {
+	p := filepath.Join(path, AWSFile)
+	return &AWSProvider{url: AWSGeoUrl, fileName: p, expire: AWSExpired}
+}
+
+func (p *AWSProvider) MaybeUpdate() error {
+	geoLock.Lock()
+	defer geoLock.Unlock()
+	if !p.NeedUpdate() && p.Info != nil {
+		return errors.New("no need to update aws file")
+	}
+	if p.NeedUpdate() {
+		tmpFile, err := ioutil.TempFile("", p.fileName+"-*")
+		if err != nil {
+			return err
+		}
+		err = os.Rename(tmpFile.Name(), p.fileName)
```

Comment:
You are again removing old db before upgrade success. If upgrade fails, old db will not longer be available. Same for other providers.